### PR TITLE
Add AlignmentSet Scatter by ZMW

### PR DESCRIPTION
**Pull request by evolvedmicrobe**
This change adds a new alignmentset scatter task, which in contrast to other AlignmentSet scatter gather tasks, such as "scatter_alignments_reference.py" and "scatter_alignments_reference_basemods.py" scatters an alignmentset by ZMW rather than by sections of the contig they align to.

Scattering by ZMWs is an appropriate mechanisms for our HMM model fitting, where the critical question is what are the error rates over a ZMW, and not over a section of reads that align to a reference.

To enable this change, I added a new python file, "scatter_alignments_reference_arrow.py", which shows features of both code re-use and code duplication.  This file was motivated by the files
"scatter_alignments_reference.py" and "scatter_alignments_reference_basemods.py", which in my summation appear to be files that are only distinguished by their task names, which I
assumed was a workaround for pbsmrtpipe issue #28.  Like "scatter_alignments_reference_basemods.py", the new file uses the parser contract in "scatter_alignments_reference.py" so re-uses code, but duplicates some of the methods in it so as to make minor changes to the scattering method and enable it to work with ZMW ranges instead of contig ranges.

This change is expected to enable enzymology while having no negative effect on production code.  A unit test has been added to ensure it works properly and will continue to do so going forward.

It should be noted that although splitting by ZMWs works very well and without error, I have some concern the designers of the API never intended to allow this use case.  In particular, there is a function `_split_zmws(...)` in `DataSetIO.py` that  can throw an exception `raise TypeError("Only ReadSets may be split by ZMWs")`, which seems like it would be an exception this may encounter, but in practice is never reached by this code and so was not a concern.

For review by @mdsmith and @mpkocher, CC: @dalexander

### CODE REVIEW

**Comment by dalexander** [comment]
AFAIK, the use cases originally imagined for splitting-by-zmw was just mapping of subreadsets .  But this is an important and real use case for splitting-by-zmws of alignmentsets.

**Comment by natechols** [comment]
Nigel, this repo has permanently moved to Bitbucket - sorry for not announcing this sooner.

**Comment by natechols** [comment]
BTW, `AlignmentSet` is a subclass of `ReadSet`, that `TypeError` is just to guard against someone trying to call `_split_zmws` on a non-BAM dataset type.

File: pbcoretools/chunking/chunk_utils.py [hunk 1]
```diff
@@ -253,24 +253,26 @@ def write_contigset_chunks_to_file(chunk_file, dataset_path, max_total_chunks, d
...
                                   max_total_nchunks, chunk_key, dir_name,
-                                  base_name, ext):
+                                  base_name, ext, byZMW):
     dset = AlignmentSet(alignmentset_path, strict=True)
-    dset_chunks = dset.split(contigs=True, maxChunks=max_total_nchunks,
-                             breakContigs=True)
-
+    if byZMW:
+        dset_chunks = dset.split(zmws=True, maxChunks=max_total_nchunks,
+                                 breakContigs=True)
```
Comment (mdsmith) [inline pbcoretools/chunking/chunk_utils.py]: It isn't obvious from the method docstring, but you don't need the breakContigs=True here.

**Comment by mdsmith** [comment]
This looks good to me. We sort the unique <moviename>_<holenumber> keys before breaking them up into ranges. I'm glad it is tested, however. I'll look at adding some unit tests as well. As for style: I try to match the surrounding code and would prefer to see by_zmw.

**Pull request closed without merging**